File: 025. K 个一组翻转链表.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(0)  # type: ignore
        dummy.next = head

        def reverse(pre, k):
            """反转 pre 后面的 k 个结点"""
            # 当 k =3 时，pre -> 1 -> 2 -> 3 -> 4 ===> pre -> 3 -> 2 -> 1 -> 4
            head = pre.next
            for i in range(k - 1):
                curr = head.next
                head.next = curr.next
                curr.next = pre.next
                pre.next = curr
            return head  # 反转后的尾节点

        def judge(node, k):
            """判断node后面是否至少还有 k 个结点"""
            temp = node
            count = 0
            while temp.next and count < k:
                temp = temp.next
                count += 1
            return count == k

        curr = dummy
        while judge(curr, k):
            curr = reverse(curr, k)
        return dummy.next

File: 025. K 个一组翻转链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	check := func(node *ListNode) bool {
		// 检查 node 后面有没有 k 个节点
		curr := node
		for i := 0; i < k; i++ {
			if curr.Next != nil {
				curr = curr.Next
			} else {
				return false
			}
		}
        return true
	}

	reverse := func(pre *ListNode) *ListNode {
		// 反转 pre 后面的 k 个节点
		curr := pre.Next
		futureTail := pre.Next
		for i := 0; i < k; i++ {
			n := curr.Next
			curr.Next = pre.Next
			pre.Next = curr
			curr = n
		}
		futureTail.Next = curr
		return futureTail
	}

	dummy := &ListNode{
		Next: head,
	}
	curr := dummy
	for check(curr) {
		curr = reverse(curr)
	}
	return dummy.Next
}